### PR TITLE
MGMT-14277: Extract recovery BaseISO

### PR DESCRIPTION
**Pull request by nmagnezi**

### CODE REVIEW

**Comment by nmagnezi** [comment]
Works okay when used as a part of a container image:
bash
```
[root@rdu-infra-edge-03 openshift-appliance]# make run LOG_LEVEL=debug ASSETS=/home/nmagnezi/go/src/github.com/openshift-appliance/container_assets/
podman run --rm -it \
	-v /home/nmagnezi/go/src/github.com/openshift-appliance/container_assets/:/assets:Z \
	--privileged \
	quay.io/derez/openshift-appliance:latest build --log-level debug
DEBUG Fetching Env Config...
DEBUG Loading Env Config...
DEBUG Using Env Config loaded from state file
DEBUG Reusing previously-fetched Env Config
DEBUG Fetching Appliance disk image...
DEBUG Loading Appliance disk image...
DEBUG   Loading Env Config...
DEBUG   Using Env Config loaded from state file
DEBUG   Loading Appliance Config...
DEBUG   Using Appliance Config loaded from target directory
DEBUG   Loading Appliance Recovery ISO...
DEBUG     Loading Env Config...
DEBUG     Loading Base ISO (CoreOS)...
DEBUG       Loading Env Config...
DEBUG       Loading Appliance Config...
DEBUG   Loading Base disk image (CoreOS)...
DEBUG     Loading Env Config...
DEBUG     Loading Appliance Config...
DEBUG   Fetching Env Config...
DEBUG   Reusing previously-fetched Env Config
DEBUG   Fetching Appliance Config...
DEBUG   Reusing previously-fetched Appliance Config
DEBUG   Fetching Appliance Recovery ISO...
DEBUG     Fetching Env Config...
DEBUG     Reusing previously-fetched Env Config
DEBUG     Fetching Base ISO (CoreOS)...
DEBUG       Fetching Env Config...
DEBUG       Reusing previously-fetched Env Config
DEBUG       Fetching Appliance Config...
DEBUG       Reusing previously-fetched Appliance Config
DEBUG     Generating Base ISO (CoreOS)...
\ Downloading CoreOS ISO...DEBUG Fetching image from OCP release (oc adm release info --image-for=machine-os-images --insecure=true quay.io/openshift-release-dev/ocp-release:4.12.0-x86_64)
\ Downloading CoreOS ISO...DEBUG extracting coreos/coreos-x86_64.iso to assets/cache, oc image extract --path coreos/coreos-x86_64.iso:assets/cache --confirm quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d4b0cfad345fbb97c2d092268dc53cdec17c5e3115212fefcfb58c7ac4652717
INFO Successfully downloaded CoreOS ISO
DEBUG   Generating Appliance Recovery ISO...
| Generating recovery ISO...DEBUG initDisk(): start
DEBUG initDisk(): regular file
DEBUG trying fat32
DEBUG fat32 failed: error reading MS-DOS Boot Sector: could not read FAT32 BIOS Parameter Block from boot sector: could not read embedded DOS 3.31 BPB: error reading embedded DOS 2.0 BPB: invalid sector size 37008 provided in DOS 2.0 BPB. Must be 512
DEBUG trying iso9660 with physical block size 0
INFO Successfully generated recovery ISO
DEBUG   Fetching Base disk image (CoreOS)...
DEBUG     Fetching Env Config...
DEBUG     Reusing previously-fetched Env Config
DEBUG     Fetching Appliance Config...
DEBUG     Reusing previously-fetched Appliance Config
DEBUG   Generating Base disk image (CoreOS)...
INFO Successfully downloaded appliance base disk image
DEBUG Generating Appliance disk image...
DEBUG Rendering guestfish/user.cfg.template
| Generating appliance disk image...DEBUG Rendering guestfish/guestfish.sh.template
DEBUG Running guestfish script
INFO Successfully generated appliance disk image
FATAL failed to fetch Appliance disk image: failed to generate asset "Appliance disk image": guestfish script failure: Failed to execute cmd (assets/temp/guestfish/guestfish.sh): assets/temp/guestfish/guestfish.sh:5: libguestfs: error: assets/cache/recovery.iso: No such file or directory
FATAL : exit status 1
make: *** [Makefile:52: run] Error 1
```

It is expected to fail afterward.
The fat32 errors are not an issue, see: https://bugzilla.redhat.com/show_bug.cgi?id=2089678

**Comment by nmagnezi** [comment]
fails with a direct binary:
first attempt:
bash
```
time="2023-04-16T08:07:58-04:00" level=debug msg="Fetching Appliance disk image..."
time="2023-04-16T08:07:58-04:00" level=debug msg="Loading Appliance disk image..."
time="2023-04-16T08:07:58-04:00" level=debug msg="  Loading Env Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="  Using Env Config loaded from state file"
time="2023-04-16T08:07:58-04:00" level=debug msg="  Loading Appliance Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="  Using Appliance Config loaded from target directory"
time="2023-04-16T08:07:58-04:00" level=debug msg="  Loading Appliance Recovery ISO..."
time="2023-04-16T08:07:58-04:00" level=debug msg="    Loading Env Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="    Loading Base ISO (CoreOS)..."
time="2023-04-16T08:07:58-04:00" level=debug msg="      Loading Env Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="      Loading Appliance Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="  Loading Base disk image (CoreOS)..."
time="2023-04-16T08:07:58-04:00" level=debug msg="    Loading Env Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="    Loading Appliance Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="  Fetching Env Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="  Reusing previously-fetched Env Config"
time="2023-04-16T08:07:58-04:00" level=debug msg="  Fetching Appliance Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="  Reusing previously-fetched Appliance Config"
time="2023-04-16T08:07:58-04:00" level=debug msg="  Fetching Appliance Recovery ISO..."
time="2023-04-16T08:07:58-04:00" level=debug msg="    Fetching Env Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="    Reusing previously-fetched Env Config"
time="2023-04-16T08:07:58-04:00" level=debug msg="    Fetching Base ISO (CoreOS)..."
time="2023-04-16T08:07:58-04:00" level=debug msg="      Fetching Env Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="      Reusing previously-fetched Env Config"
time="2023-04-16T08:07:58-04:00" level=debug msg="      Fetching Appliance Config..."
time="2023-04-16T08:07:58-04:00" level=debug msg="      Reusing previously-fetched Appliance Config"
time="2023-04-16T08:07:58-04:00" level=debug msg="    Generating Base ISO (CoreOS)..."
time="2023-04-16T08:07:59-04:00" level=debug msg="Fetching image from OCP release (oc adm release info --image-for=machine-os-images --insecure=true quay.io/openshift-release-dev/ocp-release:4.12.0-x86_64)"
time="2023-04-16T08:08:00-04:00" level=debug msg="extracting coreos/coreos-x86_64.iso to assets/cache, oc image extract --path coreos/coreos-x86_64.iso:assets/cache --confirm quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d4b0cfad345fbb97c2d092268dc53cdec17c5e3115212fefcfb58c7ac4652717"
time="2023-04-16T08:08:17-04:00" level=info msg="Successfully downloaded CoreOS ISO"
time="2023-04-16T08:08:17-04:00" level=fatal msg="failed to fetch Appliance disk image: failed to fetch dependency of \"Appliance disk image\": failed to fetch dependency of \"Appliance Recovery ISO\": failed to generate asset \"Base ISO (CoreOS)\": Failed to execute cmd (/bin/coreos-installer iso kargs show assets/cache/coreos-x86_64.iso): Error: No karg embed areas found; old or corrupted CoreOS ISO image.\n: exit status 1"
```
second attempt (no changes):
bash
```
[root@rdu-infra-edge-03 openshift-appliance]# ./build/openshift-appliance build --dir ./assets/ --log-level=debug
DEBUG Fetching Env Config...
DEBUG Loading Env Config...
DEBUG Using Env Config loaded from state file
DEBUG Reusing previously-fetched Env Config
DEBUG Fetching Appliance disk image...
DEBUG Loading Appliance disk image...
DEBUG   Loading Env Config...
DEBUG   Using Env Config loaded from state file
DEBUG   Loading Appliance Config...
DEBUG   Using Appliance Config loaded from target directory
DEBUG   Loading Appliance Recovery ISO...
DEBUG     Loading Env Config...
DEBUG     Loading Base ISO (CoreOS)...
DEBUG       Loading Env Config...
DEBUG       Loading Appliance Config...
DEBUG   Loading Base disk image (CoreOS)...
DEBUG     Loading Env Config...
DEBUG     Loading Appliance Config...
DEBUG   Fetching Env Config...
DEBUG   Reusing previously-fetched Env Config
DEBUG   Fetching Appliance Config...
DEBUG   Reusing previously-fetched Appliance Config
DEBUG   Fetching Appliance Recovery ISO...
DEBUG     Fetching Env Config...
DEBUG     Reusing previously-fetched Env Config
DEBUG     Fetching Base ISO (CoreOS)...
DEBUG       Fetching Env Config...
DEBUG       Reusing previously-fetched Env Config
DEBUG       Fetching Appliance Config...
DEBUG       Reusing previously-fetched Appliance Config
DEBUG     Generating Base ISO (CoreOS)...
INFO Reusing base ISO from cache...
DEBUG   Generating Appliance Recovery ISO...
DEBUG initDisk(): start
DEBUG initDisk(): regular file
| Nir Generating recovery ISO...DEBUG trying fat32
DEBUG fat32 failed: error reading MS-DOS Boot Sector: could not read FAT32 BIOS Parameter Block from boot sector: could not read embedded DOS 3.31 BPB: error reading embedded DOS 2.0 BPB: invalid sector size 37008 provided in DOS 2.0 BPB. Must be 512
DEBUG trying iso9660 with physical block size 0
INFO Nir Successfully generated recovery ISO
DEBUG   Fetching Base disk image (CoreOS)...
DEBUG     Fetching Env Config...
DEBUG     Reusing previously-fetched Env Config
DEBUG     Fetching Appliance Config...
DEBUG     Reusing previously-fetched Appliance Config
DEBUG   Generating Base disk image (CoreOS)...
INFO Successfully downloaded appliance base disk image
FATAL failed to fetch Appliance disk image: failed to fetch dependency of "Appliance disk image": failed to generate asset "Base disk image (CoreOS)": Failed to execute cmd (/bin/coreos-installer download -s stable -p qemu -f qcow2.xz --architecture x86_64 --decompress -C assets/cache): Read disk 13.8 MiB/634.6 MiB (2%)
FATAL Read disk 38.1 MiB/634.6 MiB (5%)
FATAL Read disk 99.7 MiB/634.6 MiB (15%)
FATAL Read disk 148.3 MiB/634.6 MiB (23%)
FATAL Read disk 162.5 MiB/634.6 MiB (25%)
FATAL Read disk 235.7 MiB/634.6 MiB (37%)
FATAL Read disk 255.7 MiB/634.6 MiB (40%)
FATAL Read disk 274.4 MiB/634.6 MiB (43%)
FATAL Read disk 294.9 MiB/634.6 MiB (46%)
FATAL Read disk 316.6 MiB/634.6 MiB (49%)
FATAL Read disk 335.7 MiB/634.6 MiB (52%)
FATAL Read disk 364.6 MiB/634.6 MiB (57%)
FATAL Read disk 378.5 MiB/634.6 MiB (59%)
FATAL Read disk 392.7 MiB/634.6 MiB (61%)
FATAL Read disk 407.5 MiB/634.6 MiB (64%)
FATAL Read disk 429.4 MiB/634.6 MiB (67%)
FATAL Read disk 444.1 MiB/634.6 MiB (69%)
FATAL Read disk 460.0 MiB/634.6 MiB (72%)
FATAL Read disk 476.6 MiB/634.6 MiB (75%)
FATAL Read disk 492.2 MiB/634.6 MiB (77%)
FATAL Read disk 506.8 MiB/634.6 MiB (79%)
FATAL Read disk 521.3 MiB/634.6 MiB (82%)
FATAL Read disk 536.1 MiB/634.6 MiB (84%)
FATAL Read disk 580.1 MiB/634.6 MiB (91%)
FATAL Read disk 634.6 MiB/634.6 MiB (100%)
FATAL gpg: Signature made Mon 03 Apr 2023 05:45:00 PM EDT
FATAL gpg:                using RSA key ACB5EE4E831C74BB7C168D27F55AD3FB5323552A
FATAL gpg: Can't check signature: No public key
FATAL Error: decoding and writing image
FATAL
FATAL Caused by:
FATAL     GPG verification failure
FATAL : exit status 1
```

**Comment by nmagnezi** [comment]
binary worked okay in another machine in one run:
bash
```
[root@nmagnezi-baremetal02 openshift-appliance]# ./build/openshift-appliance build --dir ./assets/ --log-level=debug
DEBUG Fetching Env Config...
DEBUG Loading Env Config...
DEBUG Generating Env Config...
DEBUG Fetching Appliance disk image...
DEBUG Loading Appliance disk image...
DEBUG   Loading Env Config...
DEBUG   Using Env Config loaded from state file
DEBUG   Loading Appliance Config...
DEBUG   Using Appliance Config loaded from target directory
DEBUG   Loading Appliance Recovery ISO...
DEBUG     Loading Env Config...
DEBUG     Loading Base ISO (CoreOS)...
DEBUG       Loading Env Config...
DEBUG       Loading Appliance Config...
DEBUG   Loading Base disk image (CoreOS)...
DEBUG     Loading Env Config...
DEBUG     Loading Appliance Config...
DEBUG   Fetching Env Config...
DEBUG   Reusing previously-fetched Env Config
DEBUG   Fetching Appliance Config...
DEBUG   Reusing previously-fetched Appliance Config
DEBUG   Fetching Appliance Recovery ISO...
DEBUG     Fetching Env Config...
DEBUG     Reusing previously-fetched Env Config
DEBUG     Fetching Base ISO (CoreOS)...
DEBUG       Fetching Env Config...
DEBUG       Reusing previously-fetched Env Config
DEBUG       Fetching Appliance Config...
DEBUG       Reusing previously-fetched Appliance Config
DEBUG     Generating Base ISO (CoreOS)...
/ Downloading CoreOS ISO...DEBUG Fetching image from OCP release (oc adm release info --image-for=machine-os-images --insecure=true quay.io/openshift-release-dev/ocp-release:4.12.0-x86_64)
/ Downloading CoreOS ISO...DEBUG extracting coreos/coreos-x86_64.iso to assets/cache, oc image extract --path coreos/coreos-x86_64.iso:assets/cache --confirm quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d4b0cfad345fbb97c2d092268dc53cdec17c5e3115212fefcfb58c7ac4652717
INFO Successfully downloaded CoreOS ISO
DEBUG   Generating Appliance Recovery ISO...
| Generating recovery ISO...DEBUG initDisk(): start
DEBUG initDisk(): regular file
DEBUG trying fat32
DEBUG fat32 failed: error reading MS-DOS Boot Sector: could not read FAT32 BIOS Parameter Block from boot sector: could not read embedded DOS 3.31 BPB: error reading embedded DOS 2.0 BPB: invalid sector size 37008 provided in DOS 2.0 BPB. Must be 512
DEBUG trying iso9660 with physical block size 0
INFO Successfully generated recovery ISO
DEBUG   Fetching Base disk image (CoreOS)...
DEBUG     Fetching Env Config...
DEBUG     Reusing previously-fetched Env Config
DEBUG     Fetching Appliance Config...
DEBUG     Reusing previously-fetched Appliance Config
DEBUG   Generating Base disk image (CoreOS)...
| Downloading appliance base disk image...
INFO Successfully downloaded appliance base disk image
DEBUG Generating Appliance disk image...
DEBUG Rendering guestfish/user.cfg.template
| Generating appliance disk image...DEBUG Rendering guestfish/guestfish.sh.template
DEBUG Running guestfish script
INFO Successfully generated appliance disk image
FATAL failed to fetch Appliance disk image: failed to generate asset "Appliance disk image": guestfish script failure: Failed to execute cmd (assets/temp/guestfish/guestfish.sh): assets/temp/guestfish/guestfish.sh:5: libguestfs: error: assets/cache/recovery.iso: No such file or directory
FATAL : exit status 1
```

output:
```
[root@nmagnezi-baremetal02 cache]# tree
.
├── coreos-x86_64.iso
├── fedora-coreos-37.20230322.3.0-qemu.x86_64.qcow2
└── recovery_iso
    ├── coreos
    │   ├── features.json
    │   ├── kargs.json
    │   └── miniso.dat
    ├── EFI
    │   └── redhat
    │       └── grub.cfg
    ├── images
    │   ├── efiboot.img
    │   ├── ignition.img
    │   └── pxeboot
    │       ├── initrd.img
    │       ├── rootfs.img
    │       └── vmlinuz
    ├── isolinux
    │   ├── boot.cat
    │   ├── boot.msg
    │   ├── isolinux.bin
    │   ├── isolinux.cfg
    │   ├── ldlinux.c32
    │   ├── libcom32.c32
    │   ├── libutil.c32
    │   └── vesamenu.c32
    └── zipl.prm

7 directories, 20 files
```

File: pkg/asset/recovery/recovery_iso.go [hunk 1]
```diff
@@ -39,21 +43,36 @@ func (a *RecoveryISO) Generate(dependencies asset.Parents) error {
...
 
-	isoPath := filepath.Join(envConfig.CacheDir, recoveryIsoFilename)
-	a.File = &asset.File{Filename: isoPath}
-	a.LiveISOVersion = baseISO.LiveISOVersion
+	if err := os.MkdirAll(recoveryIsoDirPath, os.ModePerm); err != nil {
+		logrus.Errorf("Failed to create dir: %s", recoveryIsoDirPath)
+	}
+
+	if err := isoeditor.Extract(coreosIsoPath, recoveryIsoDirPath); err != nil {
+		logrus.Debugf("err is: %s", err.Error()) // remove
```
Comment (danielerez) [inline pkg/asset/recovery/recovery_iso.go:54]: Worth changing to an Error log, like in line 50.

**Pull request merged**